File: mc/rcmd/src/main/java/calebxzhou/rdi/mc/rcmd/RcmdParser.java
package calebxzhou.rdi.mc.rcmd;

import java.util.ArrayList;
import java.util.List;

public final class RcmdParser {
    private RcmdParser() {
    }

    public static List<String> tokenize(String input) throws RcmdParseException {
        List<String> tokens = new ArrayList<String>();
        if (input == null) {
            return tokens;
        }
        StringBuilder token = new StringBuilder();
        boolean quoted = false;
        char quoteChar = 0;
        boolean escaping = false;
        for (int i = 0; i < input.length(); i++) {
            char c = input.charAt(i);
            if (escaping) {
                token.append(c);
                escaping = false;
                continue;
            }
            if (c == '\\') {
                escaping = true;
                continue;
            }
            if (quoted) {
                if (c == quoteChar) {
                    quoted = false;
                } else {
                    token.append(c);
                }
                continue;
            }
            if (c == '"' || c == '\'') {
                quoted = true;
                quoteChar = c;
                continue;
            }
            if (Character.isWhitespace(c)) {
                addToken(tokens, token);
                continue;
            }
            token.append(c);
        }
        if (escaping) {
            token.append('\\');
        }
        if (quoted) {
            throw new RcmdParseException("引号未闭合");
        }
        addToken(tokens, token);
        return tokens;
    }

    private static void addToken(List<String> tokens, StringBuilder token) {
        if (token.length() == 0) {
            return;
        }
        tokens.add(token.toString());
        token.setLength(0);
    }
}

File: mc/rcmd/src/main/kotlin/calebxzhou/rdi/mc/rcmd/RcmdParser.kt
package calebxzhou.rdi.mc.rcmd

object RcmdParser {
    @JvmStatic
    @Throws(RcmdParseException::class)
    fun tokenize(input: String?): List<String> {
        if (input.isNullOrEmpty()) {
            return emptyList()
        }
        val tokens = mutableListOf<String>()
        val token = StringBuilder()
        var quoted = false
        var quoteChar = '\u0000'
        var escaping = false

        for (c in input) {
            when {
                escaping -> {
                    token.append(c)
                    escaping = false
                }
                c == '\\' -> escaping = true
                quoted && c == quoteChar -> quoted = false
                quoted -> token.append(c)
                c == '"' || c == '\'' -> {
                    quoted = true
                    quoteChar = c
                }
                c.isWhitespace() -> addToken(tokens, token)
                else -> token.append(c)
            }
        }

        if (escaping) {
            token.append('\\')
        }
        if (quoted) {
            throw RcmdParseException("引号未闭合")
        }
        addToken(tokens, token)
        return tokens
    }

    private fun addToken(tokens: MutableList<String>, token: StringBuilder) {
        if (token.isEmpty()) {
            return
        }
        tokens += token.toString()
        token.clear()
    }
}
